vicky/src/lib/database/entities/task.rs

```diff
@@ -0,0 +1,415 @@
+        self.flake_ref.args.push(flake_arg.into());
+        self
+    }
+
+    pub fn with_flake_args(mut self, args: Vec<String>) -> Self {
+        self.flake_ref.args = args;
+        self
+    }
+
+    pub fn requires_feature<S: Into<String>>(mut self, feature: S) -> Self {
+        self.features.push(feature.into());
+        self
+    }
+
+    pub fn requires_features(mut self, features: Vec<String>) -> Self {
+        self.features = features;
+        self
+    }
+
+    pub fn id(&self) -> Option<Uuid> {
+        self.id
+    }
+
+    pub fn display_name(&self) -> &Option<String> {
+        &self.display_name
+    }
+
+    pub fn status(&self) -> &TaskStatus {
+        &self.status
+    }
+
+    pub fn locks(&self) -> &Vec<Lock> {
+        &self.locks
+    }
+
+    pub fn flake_ref(&self) -> &FlakeRef {
+        &self.flake_ref
+    }
+
+    pub fn features(&self) -> &Vec<String> {
+        &self.features
+    }
+
+    pub fn build(self) -> Task {
+        Task {
+            id: self.id.unwrap_or_else(Uuid::new_v4),
+            display_name: self.display_name.unwrap_or_else(|| "Task".to_string()),
+            features: self.features,
+            status: self.status,
+            locks: self.locks,
+            flake_ref: self.flake_ref,
+        }
+    }
+}
+
+// this was on purpose because these macro-generated entity types
+// mess up the whole namespace and HAVE to be scoped
+pub mod db_impl {
+    use crate::database::entities::lock::Lock;
+    use crate::database::entities::task::{Task, TaskResult, TaskStatus};
+    use crate::database::entities::FlakeRef;
+    use crate::errors::VickyError;
+    use diesel::{
+        insert_into, AsChangeset, ExpressionMethods, Identifiable, Insertable, QueryDsl, Queryable,
+        RunQueryDsl, Selectable,
+    };
+    use std::collections::HashMap;
+    use std::fmt::Display;
+    use uuid::Uuid;
+    // these here are evil >:(
+    use crate::database::schema::locks;
+    use crate::database::schema::tasks;
+    use itertools::Itertools;
+    use rocket_sync_db_pools::database;
+
+    #[derive(Insertable, Queryable, AsChangeset, Debug)]
+    #[diesel(table_name = tasks)]
+    struct DbTask {
+        pub id: Uuid,
+        pub display_name: String,
+        pub status: String,
+        pub features: String,
+        pub flake_ref_uri: String,
+        pub flake_ref_args: String,
+    }
+
+    impl Display for TaskStatus {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            let str = match self {
+                TaskStatus::NEW => "NEW",
+                TaskStatus::RUNNING => "RUNNING",
+                TaskStatus::FINISHED(r) => match r {
+                    TaskResult::SUCCESS => "FINISHED::SUCCESS",
+                    TaskResult::ERROR => "FINISHED::ERROR",
+                },
+            };
+            write!(f, "{}", str)
+        }
+    }
+
+    impl From<&str> for TaskStatus {
+        fn from(str: &str) -> TaskStatus {
```

Comment:
We might want to use TryFrom here and fail explicitly, if we see a str, which is not associated with a task state and not default to `NEW`

Reply:
Good idea.